crates/chain/src/tx_graph.rs

```diff
@@ -719,8 +725,40 @@ impl<A: Anchor> TxGraph<A> {
 
     /// Inserts the given `seen_at` for `txid` into [`TxGraph`].
     ///
-    /// Note that [`TxGraph`] only keeps track of the latest `seen_at`.
+    /// Note that [`TxGraph`] keeps track of the latest `seen_at` and the earliest `seen_at`
     pub fn insert_seen_at(&mut self, txid: Txid, seen_at: u64) -> ChangeSet<A> {
+        let mut changeset_first_seen = self.update_first_seen(txid, seen_at);
+        let changeset_last_seen = self.update_last_seen(txid, seen_at);
+        changeset_first_seen.merge(changeset_last_seen);
+        changeset_first_seen
+    }
+
+    /// Updates `first_seen` given a new `seen_at`
+    fn update_first_seen(&mut self, txid: Txid, seen_at: u64) -> ChangeSet<A> {
+        let is_changed = match self.first_seen.entry(txid) {
+            hash_map::Entry::Occupied(mut e) => {
+                let first_seen = e.get_mut();
+                let change = *first_seen > seen_at;
+                if change {
+                    *first_seen = seen_at;
+                }
+                change
+            }
+            hash_map::Entry::Vacant(e) => {
+                e.insert(seen_at);
+                true
+            }
+        };
+
+        let mut changeset = ChangeSet::<A>::default();
+        if is_changed {
+            changeset.first_seen.insert(txid, seen_at);
+        }
+        changeset
+    }
+
+    /// Updates `last_seen` given a new `seen_at`
+    pub fn update_last_seen(&mut self, txid: Txid, seen_at: u64) -> ChangeSet<A> {
```

Comment:
Please make this private. We always want to update both `last-seen` and `first-seen` whenever a timestamp is introduced.